blockchain/utxocache.go

```diff
@@ -0,0 +1,790 @@
+	}
+
+	// Flush if the max size of the cache has been reached.
+	if c.totalSize() >= c.maxSize {
+		return true
+	}
+
+	// Flush if the periodic flush duration has been reached.
+	now := time.Now()
+	duration := now.Sub(c.lastFlushTime)
+	return duration >= time.Minute*periodicFlushMins
+}
+
+// flush commits all modified entries to the database and conditionally evicts
+// entries.
+//
+// Entries that are nil or spent are always evicted since they are
+// unlikely to be accessed again.  Additionally, if the cache has reached its
+// maximum size, entries are evicted based on the height of the block that they
+// are contained in.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) flush(bestHash *chainhash.Hash, bestHeight uint32, logFlush bool) error {
+	// If the maximum allowed size of the cache has been reached, determine the
+	// eviction height.
+	var evictionHeight uint32
+	memUsageMiB := float64(c.totalSize()) / 1024 / 1024
+	maxMemUsageMiB := float64(c.maxSize) / 1024 / 1024
+	if memUsageMiB >= maxMemUsageMiB {
+		evictionHeight = c.calcEvictionHeight(bestHeight)
+	}
+
+	// Log that a flush is starting and indicate the current memory usage, hit
+	// ratio, and eviction height.
+	var hitRatio float64
+	var evictionLog string
+	if logFlush {
+		memUsagePercent := memUsageMiB / maxMemUsageMiB * 100
+		hitRatio = c.hitRatio()
+		if evictionHeight != 0 {
+			evictionLog = fmt.Sprintf(", eviction height: %d", evictionHeight)
+		}
+		log.Infof("UTXO cache flush starting (%d entries, %.2f MiB (%.2f%%), "+
+			"%.2f%% hit ratio, height: %d%s)", len(c.entries), memUsageMiB,
+			memUsagePercent, hitRatio, bestHeight, evictionLog)
+	}
+
+	// Flush the entries in the cache to the database and update the utxo set
+	// state in the database.
+	//
+	// It is important that the utxo set state is always updated in the same
+	// database transaction as the utxo set itself so that it is always in sync.
+	err := c.db.Update(func(dbTx database.Tx) error {
+		for outpoint, entry := range c.entries {
+			// Write the entry to the database.
+			err := dbPutUtxoEntry(dbTx, outpoint, entry)
+			if err != nil {
+				return err
+			}
+
+			// Conditionally evict the entry from the cache.  Entries that are nil or
+			// spent are always evicted since they are unlikely to be accessed again.
+			// Additionally, entries that are contained in a block with a height less
+			// than the eviction height are evicted.
+			if entry == nil || entry.IsSpent() ||
+				entry.BlockHeight() < int64(evictionHeight) {
+
+				// Remove the entry from the cache.
+				delete(c.entries, outpoint)
+
+				// Update the total entry size of the cache.
+				if entry != nil {
+					c.totalEntrySize -= entry.size()
+				}
+
+				continue
+			}
+
+			// If the entry wasn't removed from the cache, clear the modified and
+			// fresh flags since it has been updated in the database.
+			entry.state &^= utxoStateModified
+			entry.state &^= utxoStateFresh
```

Comment:
I thought about this as well, but as you noted, I was thinking that if database transactions are erroring out, there is likely a larger and unrecoverable underlying issue.

Historically, have there been cases where leveldb throws an unexpected error but then quickly recovers?  If this can occur, it may make sense to do something generically, like automatically retry the transaction N number of times with exponential backoffs and then panic if it still fails.

In any case, I will update as suggested.  I'm thinking it may actually be more efficient to just loop the entries map again after the db tx to handle the eviction rather than tracking what needs to be deleted/modified in dynamically allocated maps to avoid the memory usage/inefficiency but I'll see what works best.

Reply:
> Historically, have there been cases where leveldb throws an unexpected error but then quickly recovers? 

It is pretty rare, but it does happen when hardware is failing.  We've had several reports over the years where people see strange corruption errors and things of the like which invariably end up being due to failing memory or storage devices.

>  I'm thinking it may actually be more efficient to just loop the entries map again ...

Yeah, I think just doing a double loop makes sense here.  It avoids more memory usage and cleanly fixes the potential rollback issues in exchange for slightly more CPU usage, but it happens infrequently, so it's a good tradeoff.